Add user profiles closes #165

alexjgriffith opened the pull request:
## Changes
- Adds user profiles 
- also provides functions `mastodon-profile--get-following` and `mastodon-profile--get-followers`
- added keybinding U for opening user profiles
- autoloads `mastodon-profile--get-next-author`

## Notes
Should add tests for a few functions. But most of this builds directly on `mastodon-tl` functionality.

alexjgriffith committed:
Add user profiles

- added keybinding U for opening user profiles
- autoloads 'mastodon-profile--get-next-author

hdurer commented on lisp/mastodon-profile.el:
```diff
@@ -0,0 +1,154 @@
+         (json (mastodon-http--get-json url)))
+    (with-output-to-temp-buffer buffer
+      (switch-to-buffer buffer)
+      (mastodon-mode)
```
If you move this to the very end, once you have created all the contents then you can get rid of the `inhibit-read-only` setting below.

alexjgriffith replied:
If I don't put `mastodon-mode` up top I get a weird error when requesting more toots stating that `mastodon-tl--buffer-spec` is not defined.

hdurer commented on lisp/mastodon-profile.el:
```diff
@@ -0,0 +1,154 @@
+                          update-function
+                          ,'mastodon-tl--timeline json))
+      (let ((inhibit-read-only t))
+        "\n"
```
You need to wrap all this until the `(mastodon-tl--timeline json)` below in an `insert` call.
On my machine this doesn't make it into the buffer otherwise.

hdurer commented on lisp/mastodon-profile.el:
```diff
@@ -0,0 +1,154 @@
+        "\n"
+        (mastodon-profile--image-from-status account)
+        "\n"
+        (propertize (mastodon-profile--field status 'display_name
```
This is bracketed incorrectly.  `(mastodon-profile--field status 'display_name)` is the inner expression otherwise I get an error and there are no properties to set.

hdurer commented on lisp/mastodon-profile.el:
```diff
@@ -0,0 +1,154 @@
+  (mastodon-media--inline-images))
+
+
+(defun mastodon-profile--get-following ()
```
This and `mastodon-profile--followers` aren't hooked up yet?

alexjgriffith replied:
They work, but i could not think of an intuitive key binding.

We can brainstorm and give them a keybinding before the 0.8.0 release.

hdurer replied:
Sounds good.

hdurer commented on lisp/mastodon-profile.el:
```diff
@@ -0,0 +1,154 @@
+    (unless (equal url "/avatars/original/missing.png")
+      (mastodon-media--get-media-link-rendering url))))
+
+(defun mastodon-profile--field (status field)
```
Neither the function name nor the docstring gives me an idea of what this does.

I think you should work with the account json object and not the toot.  Then you can lose the inner assoc and this becomes a simple concatenation of `cdr` and `assoc` which doesn't really add any abstraction so we could get rid of it altogether.

hdurer commented on lisp/mastodon-profile.el:
```diff
@@ -0,0 +1,154 @@
+            (insert (propertize (mastodon-tl--byline-author
+                                 (list (append (list 'account) toot)))
+                                'byline  't
+                                'toot-id (cdr(assoc 'id toot)) 'toot-json toot)
```
Nit: Missing space after `cdr` here and twice more below.

hdurer reviewed (commented):
I really like this!

A few changes had to be made to make this work for me.

Thinking ahead how this can work given a generic user handle (when we process the "@userid" links in the body) I suspect we should work off account objects and not full toots?

Nothing in `mastodon-profile--make-author-buffer` really needs the full status but only wants data from the account sub-object.  (for that to be true, we need to make the second arg to `mastodon-tl--render-text` optional - it won't be needed when rendering a user's self description I suspect).

alexjgriffith committed:
User profiles now use accounts rather than statuses.

- renamed mastodon-profile--next-toot-author to mastodon-profile--get-toot-author
- mastodon-tl--render-text no longer parses text if toot is nil
- mastodon-profile--field is now mastodon-profile--account-field and it now accepts an account rather than a status
- Added info in readme
- Added vimish keybindings to tab and stab in mastodon-mode to compliment jk nav
- Other small fixes to mastodon-profile.el

alexjgriffith committed:
Merge branch 'develop' into user-profiles

hdurer reviewed (approved):
Cool.  This is quite exciting!
